### ISSUE TITLE
Spacer hangs/errors in proof generation without slicing/inlining

### ISSUE THREAD

**Issue by leonardoalt** [Z3Prover/z3#5363]
Sample: https://gist.github.com/leonardoalt/8a9181ba0b0f68d02b49bebf12df0deb
Version: 4.8.11 pre-release

Proving takes only 0.13s (without `print-certificate`), but the generation step takes several minutes and then fails:

```
time z3 out1.smt2 rewriter.pull_cheap_ite=true fp.spacer.q3.use_qgen=true fp.spacer.mbqi=false fp.spacer.ground_pobs=false fp.xform.slice=false fp.xform.inline_linear=false fp.xform.inline_eager=false
sat
(error "line 1113 column 47: spacer: could not validate a proof step")
z3 out1.smt2 rewriter.pull_cheap_ite=true fp.spacer.q3.use_qgen=true       492.30s user 0.73s system 99% cpu 8:13.64 total
```
Removing the options that disable slicing and inlining (last 3) makes it work, solving in 0.05s and producing the proof in 2s.

**Comment by NikolajBjorner** [comment]
It hits some limitations on dealing with arrays in Spacer. It bails out when a post-hoc solver call returns unknown.

1. Spacer relies on solving SMT queries to produce a counter-example trace. Information sufficient to produce the trace is not stored during the search. It presumably could be re-engineered to ensure this information is available without relying on post-hoc SMT calls. This is without doubt a somewhat substantial change, but worth considering in this light.
2. The formula sent to the solver leaks some functions created locally in models so the formula isn't really self-contained. It also represents constant arrays using lambdas that are not generally going to let the solver return SAT (instead it may return unknown). 

A relevant snippet from the trace reconstruction is here:

```
                              14
                              16)
                       (lambda ((x!1 Int)) (= x!1 5))
                       a!3)))
  (and (= summary_constructor_10_C_103_0_0_n 0)
       (= summary_constructor_10_C_103_0_1_n 5)
       (= summary_constructor_10_C_103_0_2_n abi_type)
       (= summary_constructor_10_C_103_0_3_n
          (crypto_type ((as const (Array ecrecover_input_type Int)) 0)
                       ((as const (Array bytes_tuple Int)) 0)
                       ((as const (Array bytes_tuple Int)) 0)
                       ((as const (Array bytes_tuple Int)) 0)))
       (= summary_constructor_10_C_103_0_4_n
          (tx_type 20584
                   31597
                   17887
                   28223
                   6878
                   29533
                   (_ as-array k!2031)
                   (bytes_tuple (_ as-array k!2030) 10)
                   0
                   11
                   0
                   867
```

The emergence of  k!2031 are not directly to blame SPACER though debugging this is a bit time consuming.

**Comment by leonardoalt** [comment]
Thanks for the explanation @NikolajBjorner .

1) The reason I'm disabling inlining and slicing in the first place is to get complete counterexamples, since they do look "sliced" when there's heavy inlining/slicing involved, and I can't relate it back to the program. If the cex trace is reengineered to contain enough information to build the cex, could this be combined with something that also knows how to go back to the original problem, even in the presence of inlining/slicing?
2) The lambda representation for constant arrays started recently for me, when I added an `(Array Int Bool)` to my encoding. I haven't seen the array lambda for other sorts.

I understand this is a pretty complex case and that this query's cex generation is pushing a bit too far, so please feel free to rename/move/close the issue.

**Comment by NikolajBjorner** [comment]
Closing per fixes in d5c6abe14d3edc18667e5828144816ec756289fb

The status is that the example now goes through. 
Counter-example trace could use an overhaul to be much more efficient (and not brittle).

The lambdas come from Boolean arrays and how they get simplified into conjunctions instead of select/store. 
They don't influence solvability so I am not trying to change upstream introduction of lambdas for this downstream use.

**Comment by agurfinkel** [comment]
This is something that I should address. Unfortunately, it will be some time until I will get to it.

Caching models might be expensive, both because the models have to be computed during solving and they have to be kept for all predecessor queries...
